### PR TITLE
[wheel][cpp] recover cpp extra

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?
- We will release C++ API in 1.7.0. We need to recover cpp extra and support installing ray-cpp by the way "ray install ray[cpp]".
- Please merge this PR after the issue #18538 addressed.

### CODE REVIEW

File: python/setup.py [hunk 1]
```diff
@@ -198,7 +199,7 @@ def get_packages(self):
         ],
     }
 
-    if os.getenv("RAY_EXTRA_CPP") == "1":
+    if RAY_EXTRA_CPP:
```
Comment (krfricke) [inline python/setup.py:202]: IMO we should check here if the `setup_spec.version` is a proper version (like 1.6.0) or the development version. We will not upload development cpp wheels to pypi. Thus, when installing a dev version from a specific commit, installing `ray[all]` afterwards will always downgrade ray to the latest release.

**Review by krfricke — CHANGES_REQUESTED** [review]
We should still have a way to exclude the extra requirement (e.g. `os.getenv("RAY_EXTRA_CPP", "1") == "1"`) so we can not require this in test settings.

Also, we have to be careful with development wheels, that will be easily downgraded with subsequent `pip install ray[all]` commands.

File: python/setup.py (hunk 1 above)
Comment (krfricke) [reply python/setup.py:202]: So to be clear, if `setup_spec.version == "2.0.0.dev0"` then we want to not require a specific `ray_cpp` version or default to not including `ray_cpp` in extras

Comment (SongGuyang) [reply python/setup.py:202]: Maybe we will rename "2.0.0.dev0" to "3.0.0.dev0" later?  Decide the logic by `if "dev" in setup_spec.version`?

**Comment by SongGuyang** [comment]
@krfricke Can you review again? I added an "if" here https://github.com/ray-project/ray/pull/18597/files#diff-eb8b42d9346d0a5d371facf21a8bfa2d16fb49e213ae7c21f03863accebe0fcfR119

**Review by krfricke — APPROVED** [review]
Thanks for addressing, looks good to me

**Comment by SongGuyang** [comment]
@edoakes @ericl @richardliaw Anyone can help approve current PR? A code owner needed.

**Review by edoakes — APPROVED** [review]

**Pull request merged**